### PR TITLE
Dispose support

### PR DESCRIPTION
**Pull request by Karnah**
**Problem**
* 'Dispose' method didn't cleaning all references to extension object
* `LocExtension` is using finalizator to unsubscribe from events, because it doesn't know when last dependency is removed. Also finalizating is slow operation. 
* I think virtual method is much better then Action field

**Fixed**
- Created virtual methods OnFirstTargetAdded and OnLastTargetRemoved. They help to control subscription and unsubscription. 
- Created interface for ObjectDependencyManager. Now it can notify if all references are dead. It will allows clear them when it found. Also removed throw exception, because this is normal situation now.
- Added two new unit tests and removed one.

### CODE REVIEW

**Comment by Karnah** [comment]
Sorry, my fault. ObjectDependencyManager works fine. It's just necessary to wait more time when Finalizer collect objects.

**Comment by Karnah** [comment]
About perfomance tests. For execution: set XAMLMarkupExtensions.PerformanceTests as startup, change Build to "Release" and run. I created two tests: ProvideValue and Cleanup. But second is not so good, so it's commented.

Test perfomance:
Old code (with `ObjectDependencyManager.CleanUp(this);` )
|       Method | extensionsCount | extensionTargetsCount |    Mean |    Error |   StdDev |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|------------- |---------------- |---------------------- |--------:|---------:|---------:|------------:|-----------:|----------:|----------:|
| ProvideValue |              50 |                   100 | 1.217 s | 0.0090 s | 0.0075 s | 255000.0000 | 10000.0000 | 2000.0000 |   1.12 GB |
| ProvideValue |            3000 |                     1 | 1.745 s | 0.0123 s | 0.0109 s | 497000.0000 | 18000.0000 | 3000.0000 |   2.18 GB |

New code:
|       Method | extensionsCount | extensionTargetsCount |    Mean |    Error |   StdDev |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|------------- |---------------- |---------------------- |--------:|---------:|---------:|------------:|-----------:|----------:|----------:|
| ProvideValue |              50 |                   100 | 1.249 s | 0.0245 s | 0.0381 s | 255000.0000 | 12000.0000 | 2000.0000 |   1.12 GB |
| ProvideValue |            3000 |                     1 | 1.716 s | 0.0063 s | 0.0056 s | 497000.0000 | 17000.0000 | 3000.0000 |   2.18 GB |

So, this changes make code a bit slower at ProvideValue. But I thinks it will be a bit faster when ObjectDependencyManager will clean dead references.

**Comment by konne** [comment]
@Karnah sorry for this long silence. I had bunch of other things and now I have again time for this topics.
Please can you first split this into different PRs
1. All tests -> I will just merge them
2. All Hashset changes
3. Action -> Virtual
4. rest

I really like to understand why your new solution is slower.

Okay for you?

**Comment by Karnah** [comment]
@konne
There is a problem if I'll try separate tests from other code. It's necessary to have correct dispose to run it, because it run each many times and calculate mean of them. I think it can be solved with GC, but it is not good way.
I'll think how to sepate this to different PRs, what each PR be stable, but I'm not sure how easy it will be.

About perfomance: it becames slower only for call ProvideValue when there is not dead references. I think It happens because HashSet<> is a bit slower to add item then List<>. I'm expecting that removing dead references will be faster with HashSet<>. I'll try fix test to prove it.

**Comment by SeriousM** [comment]
Hashset is faster than a list:
https://stackoverflow.com/questions/150750/hashset-vs-list-performance#10762995

**Comment by Karnah** [comment]
@SeriousM, not always. `Add` method sould be faster for `List<>` than `HashSet<>`. Also for (as at your link said) small collections - `List<>` is faster too. I hope this changes can greatly improve perfomance for big collections (which I have) but I'd like to have prove of this.

**Comment by SeriousM** [comment]
> Also for (as at your link said) small collections

>  hope this changes can greatly improve perfomance for big collections (which I have)

How does your statements fit together? I seek for performance for large collections and ask to use `List` which is fast for collections of five items (link).

**Comment by Karnah** [comment]
@SeriousM, I'm so sorry, is was a hard day. I decided, what you commented decreasing performance and it is incorrect situation.
Yes, it should be faster and I'd like to check how much.

**Comment by Karnah** [comment]
@konne, for 1-3 I created separated independent PRs. Other changes depend of this PR's, so I'll create it after merge. Also I'd like to create unit tests for some classes (for example `ObjectDependencyManager`). What do you prefer - XUnit or NUnit? I've worked with both, but I usually use XUnit.

**Comment by konne** [comment]
> @konne, for 1-3 I created separated independent PRs.
> Other changes depend of this PR's, so I'll create it after merge.

great, I already merge 2 of your PRs.

>Also I'd like to create unit tests for some classes (for example `ObjectDependencyManager`).
>What do you prefer - XUnit or NUnit? I've worked with both, but I usually use XUnit.
Please use XUnit

**Comment by Karnah** [comment]
@konne, I completed work and updated description.

File: src/Base/IObjectDependency.cs [hunk 1]
```diff
@@ -11,13 +11,13 @@
     public interface IObjectDependency
```
Comment (konne) [inline src/Base/IObjectDependency.cs:11]: @Karnah 
I understand you intention to make the name of the Method more easy to understand.
I have no issue to change the comment, but please let the Method names, even if the
name is not perfect. This is a public Interface an even if nearly nobody will use it I don't
like to change it, for this small reason.

**Review by konne — CHANGES_REQUESTED** [review]
@Karnah please take care of this small change the rest is okay

File: src/Base/IObjectDependency.cs (hunk 1 above)
Comment (Karnah) [reply src/Base/IObjectDependency.cs:11]: Sorry, I could not understand what you mean. Should I return names OnReferencesRemoved and OnAllReferencesRemoved? Or use something else?

**Review by konne — APPROVED** [review]
@Karnah please ignore my comment, everything is okay.

**Pull request merged**